### ISSUE TITLE
Change default dropout rate

### ISSUE THREAD

**Issue by daisyron** [davisking/dlib#991]
Hi,

I would like to change the default (0.5) drop out rate in the dropout layer.
I found a past in which this question was asked (Issue #529) and tried to implement the advise there without a success.

Actually I tried to assign an dropout layer with a different drop out rate, using the following code:

`layer<2>(net) = dropout<layer<3>(net)>(0.2);`

Where is my network layer 2 is a dropout layer (assuming layer<3> is the subnet of layer<2>).

This code does not compile, giving me the following error:

cnnNetworkTrainer.cpp: In function ‘int main(int, char**)’:
cnnNetworkTrainer.cpp:565:39: error: call to non-constexpr function ‘typename dlib::impl::layer_helper<i, T>::type& dlib::layer(T&) [with unsigned int i = 3u; T = dlib::add_loss_layer<dlib::loss_multiclass_log_, dlib::add_layer<dlib::softmax_, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::fc_<2ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::fc_<20ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 5l, 5l, 1, 1, 2, 2>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 7l, 7l, 1, 1, 3, 3>, dlib::add_layer<dlib::dropout_, dlib::input_general_image_sized<float, 3ul, 90ul, 45ul>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void> >; typename dlib::impl::layer_helper<i, T>::type = dlib::add_layer<dlib::fc_<2ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::fc_<20ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 5l, 5l, 1, 1, 2, 2>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 7l, 7l, 1, 1, 3, 3>, dlib::add_layer<dlib::dropout_, dlib::input_general_image_sized<float, 3ul, 90ul, 45ul>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>]’
   layer<2>(net) = dropout<layer<3>(net)>(0.2);
                                       ^
cnnNetworkTrainer.cpp:565:40: error: type/value mismatch at argument 1 in template parameter list for ‘template<class SUBNET> using dropout = dlib::add_layer<dlib::dropout_, SUBNET>’
   layer<2>(net) = dropout<layer<3>(net)>(0.2);
                                        ^
cnnNetworkTrainer.cpp:565:40: error:   expected a type, got ‘dlib::layer<3u, dlib::add_loss_layer<dlib::loss_multiclass_log_, dlib::add_layer<dlib::softmax_, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::fc_<2ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::fc_<20ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 5l, 5l, 1, 1, 2, 2>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 7l, 7l, 1, 1, 3, 3>, dlib::add_layer<dlib::dropout_, dlib::input_general_image_sized<float, 3ul, 90ul, 45ul>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void> > >((* & net))’

Can someone please advise me how to do this properly?

BTW: isn't it make sense to implement a "set" method to change the default value? A "get" method is already implemented...

Thanks in advance,

Ron.

**Comment by davisking** [comment]
You need to call `.layer_details()` to get the layer details object before you mess with it.  Like, `layer<2>(net).layer_details().set_dropout_rate(whatever)`.

**Comment by daisyron** [comment]
Hi,

Thanks for your quick response...

I did the following:

`layer<2>(net).layer_details().set_dropout_rate(0.2f);`

And it does not compile. The compiler complain that the set_dropout_rate is not defined. The error I get:

nnNetworkTrainer.cpp: In function ‘int main(int, char**)’:
cnnNetworkTrainer.cpp:565:33: error: ‘dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::fc_<2ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::fc_<20ul, (dlib::fc_bias_mode)0u>, dlib::add_layer<dlib::dropout_, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<32l, 3l, 3l, 1, 1, 1, 1>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 5l, 5l, 1, 1, 2, 2>, dlib::add_layer<dlib::max_pool_<2l, 2l, 2, 2, 0, 0>, dlib::add_layer<dlib::relu_, dlib::add_layer<dlib::con_<64l, 7l, 7l, 1, 1, 3, 3>, dlib::add_layer<dlib::dropout_, dlib::input_general_image_sized<float, 3ul, 90ul, 45ul>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>, void>::layer_details_type’ has no member named ‘set_dropout_rate’
   layer<2>(net).layer_details().set_dropout_rate(0.2f);

Any comment?

Regards,

Ron.

**Comment by davisking** [comment]
Your second layer isn't a dropout layer.

Print the network to the screen with `cout << net;` so you can clearly see
the layer numbers.  Then you won't be confused about what numbers
correspond to which layers.

**Comment by daisyron** [comment]
Here is my network architecture with the layers numbers:
================================================================================
layer<0>        loss_multiclass_log
layer<1>        softmax
layer<2>        dropout  (drop_rate=0.5)
layer<3>        fc       (num_outputs=2) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<4>        relu
layer<5>        fc       (num_outputs=20) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<6>        dropout  (drop_rate=0.5)
layer<7>        max_pool (nr=2, nc=2, stride_y=2, stride_x=2, padding_y=0, padding_x=0)
layer<8>        relu
layer<9>        con      (num_filters=32, nr=3, nc=3, stride_y=1, stride_x=1, padding_y=1, padding_x=1) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<10>       max_pool (nr=2, nc=2, stride_y=2, stride_x=2, padding_y=0, padding_x=0)
layer<11>       relu
layer<12>       con      (num_filters=32, nr=3, nc=3, stride_y=1, stride_x=1, padding_y=1, padding_x=1) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<13>       max_pool (nr=2, nc=2, stride_y=2, stride_x=2, padding_y=0, padding_x=0)
layer<14>       relu
layer<15>       con      (num_filters=64, nr=5, nc=5, stride_y=1, stride_x=1, padding_y=2, padding_x=2) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<16>       max_pool (nr=2, nc=2, stride_y=2, stride_x=2, padding_y=0, padding_x=0)
layer<17>       relu
layer<18>       con      (num_filters=64, nr=7, nc=7, stride_y=1, stride_x=1, padding_y=3, padding_x=3) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0
layer<19>       dropout  (drop_rate=0.5)
layer<20>       input_general_image_sized(nr=90 nc=45 nofChannels=3 avgs=(0,0,0)
================================================================================

layer<2> is a dropout, isn't it?

Ron.

**Comment by daisyron** [comment]
Her is my network definition as well:

using cnnNetType = 
      loss_multiclass_log<
      softmax<dropout<
      fc<NOF_OUT_CLASSES,
      relu<fc<20,dropout<
      max_pool<2,2,2,2,relu<con<32,3,3,1,1,
      max_pool<2,2,2,2,relu<con<32,3,3,1,1,
      max_pool<2,2,2,2,relu<con<64,5,5,1,1,
      max_pool<2,2,2,2,relu<con<64,7,7,1,1,
      dropout<InputTensorType
      >>>>>>>>>>>>>>>>>>>>;

**Comment by davisking** [comment]
Oops, it's `set_drop_rate()`, not `set_dropout_rate()`.

**Comment by davisking** [comment]
Sorry, I'm replying way too quickly.  There is no such function.  There is only get_drop_rate.  Anyway, now that I'm actually looking at the documentation (which you should be doing too ;)), I see there is no function.  So instead just assign the layer details with a new dropout layer like  `layer<2>(net).layer_details() = dropout_(whatever)`;

**Comment by daisyron** [comment]
Thanks!!!

This solved my problem.

Regards,

Ron.
